File: backend/app/services/session_manager.py
"""
Session manager for storing and retrieving graph data.

Provides in-memory storage for processed graphs with optional persistence.
"""

import uuid
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, Optional
import networkx as nx
from app.core.config import settings
from app.services.graph_builder import GraphBuilder


class SessionManager:
    """
    Manages graph sessions in memory.

    Each uploaded and processed log file gets a unique session ID.
    The session stores the original messages, built graph, and metadata.
    """

    def __init__(self):
        self._sessions: Dict[str, Dict] = {}

    def create_session(
        self,
        messages: list,
        framework: str,
        graph_builder: GraphBuilder
    ) -> str:
        """
        Create a new session for a processed graph.

        Args:
            messages: List of parsed Message objects
            framework: Framework name used for parsing
            graph_builder: GraphBuilder instance with built graph

        Returns:
            Unique session ID
        """
        session_id = str(uuid.uuid4())[:8]  # Short UUID for convenience

        self._sessions[session_id] = {
            'id': session_id,
            'messages': messages,
            'framework': framework,
            'graph_builder': graph_builder,
            'created_at': datetime.now(),
            'last_accessed': datetime.now(),
            'metadata': {}
        }

        return session_id

    def get_session(self, session_id: str) -> Optional[Dict]:
        """
        Retrieve a session by ID.

        Args:
            session_id: The session identifier

        Returns:
            Session dictionary or None if not found
        """
        if session_id not in self._sessions:
            return None

        # Update last accessed time
        self._sessions[session_id]['last_accessed'] = datetime.now()

        return self._sessions[session_id]

    def get_graph_builder(self, session_id: str) -> Optional[GraphBuilder]:
        """
        Get the GraphBuilder for a session.

        Args:
            session_id: The session identifier

        Returns:
            GraphBuilder instance or None if not found
        """
        session = self.get_session(session_id)
        return session['graph_builder'] if session else None

    def delete_session(self, session_id: str) -> bool:
        """
        Delete a session.

        Args:
            session_id: The session identifier

        Returns:
            True if deleted, False if not found
        """
        if session_id in self._sessions:
            del self._sessions[session_id]
            return True
        return False

    def cleanup_expired_sessions(self) -> int:
        """
        Remove sessions older than SESSION_EXPIRY_HOURS.

        Returns:
            Number of sessions cleaned up
        """
        expiry_time = datetime.now() - timedelta(hours=settings.SESSION_EXPIRY_HOURS)
        expired_sessions = [
            sid for sid, session in self._sessions.items()
            if session['last_accessed'] < expiry_time
        ]

        for session_id in expired_sessions:
            del self._sessions[session_id]

        return len(expired_sessions)

    def get_session_count(self) -> int:
        """Get the number of active sessions."""
        return len(self._sessions)

    def list_sessions(self) -> list:
        """
        List all active session IDs with basic metadata.

        Returns:
            List of session info dictionaries
        """
        return [
            {
                'id': session['id'],
                'framework': session['framework'],
                'created_at': session['created_at'].isoformat(),
                'message_count': len(session['messages'])
            }
            for session in self._sessions.values()
        ]


# Global session manager instance
session_manager = SessionManager()

File: src/lib/services/session-manager.ts
/**
 * Session Manager - In-memory session storage for processed graphs
 *
 * Uses a module-level Map that persists across requests in the same server instance.
 * Note: Sessions are cleared on server restart (acceptable for dev/demo).
 */

import type { Message } from '@/lib/models/types';
import type { GraphBuilder } from './graph-builder';

export interface Session {
  id: string;
  messages: Message[];
  framework: string;
  graphBuilder: GraphBuilder;
  createdAt: Date;
  lastAccessed: Date;
  metadata: Record<string, unknown>;
}

export interface SessionInfo {
  id: string;
  framework: string;
  message_count: number;
  node_count: number;
  edge_count: number;
  created_at: string;
  last_accessed: string;
}

// Module-level storage - persists across requests in the same server instance
const sessions = new Map<string, Session>();

// Session expiry time (24 hours)
const SESSION_EXPIRY_MS = 24 * 60 * 60 * 1000;

/**
 * Generate a short unique session ID (8 characters).
 */
function generateSessionId(): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let id = '';
  for (let i = 0; i < 8; i++) {
    id += chars[Math.floor(Math.random() * chars.length)];
  }
  return id;
}

/**
 * Create a new session with parsed messages and graph.
 *
 * @param messages - Parsed message list
 * @param framework - Framework name used for parsing
 * @param graphBuilder - GraphBuilder instance with built graph
 * @returns Session ID
 */
export function createSession(
  messages: Message[],
  framework: string,
  graphBuilder: GraphBuilder
): string {
  // Generate unique session ID
  let sessionId = generateSessionId();
  while (sessions.has(sessionId)) {
    sessionId = generateSessionId();
  }

  const now = new Date();

  const session: Session = {
    id: sessionId,
    messages,
    framework,
    graphBuilder,
    createdAt: now,
    lastAccessed: now,
    metadata: {},
  };

  sessions.set(sessionId, session);

  // Clean up expired sessions in the background
  cleanupExpiredSessions();

  return sessionId;
}

/**
 * Get a session by ID.
 *
 * @param sessionId - Session ID to look up
 * @returns Session object or null if not found
 */
export function getSession(sessionId: string): Session | null {
  const session = sessions.get(sessionId);

  if (!session) {
    return null;
  }

  // Update last accessed time
  session.lastAccessed = new Date();

  return session;
}

/**
 * Get the GraphBuilder for a session.
 *
 * @param sessionId - Session ID
 * @returns GraphBuilder or null if session not found
 */
export function getGraphBuilder(sessionId: string): GraphBuilder | null {
  const session = getSession(sessionId);
  return session?.graphBuilder ?? null;
}

/**
 * Delete a session by ID.
 *
 * @param sessionId - Session ID to delete
 * @returns true if deleted, false if not found
 */
export function deleteSession(sessionId: string): boolean {
  return sessions.delete(sessionId);
}

/**
 * Get session info without the full graph data.
 *
 * @param sessionId - Session ID
 * @returns SessionInfo or null if not found
 */
export function getSessionInfo(sessionId: string): SessionInfo | null {
  const session = getSession(sessionId);

  if (!session) {
    return null;
  }

  const graph = session.graphBuilder.getGraph();

  return {
    id: session.id,
    framework: session.framework,
    message_count: session.messages.length,
    node_count: graph?.numberOfNodes() ?? 0,
    edge_count: graph?.numberOfEdges() ?? 0,
    created_at: session.createdAt.toISOString(),
    last_accessed: session.lastAccessed.toISOString(),
  };
}

/**
 * List all active sessions.
 *
 * @returns Array of session info objects
 */
export function listSessions(): SessionInfo[] {
  const result: SessionInfo[] = [];

  for (const [id, session] of sessions) {
    const graph = session.graphBuilder.getGraph();
    result.push({
      id,
      framework: session.framework,
      message_count: session.messages.length,
      node_count: graph?.numberOfNodes() ?? 0,
      edge_count: graph?.numberOfEdges() ?? 0,
      created_at: session.createdAt.toISOString(),
      last_accessed: session.lastAccessed.toISOString(),
    });
  }

  return result;
}

/**
 * Clean up expired sessions.
 *
 * @returns Number of sessions cleaned up
 */
export function cleanupExpiredSessions(): number {
  const now = Date.now();
  let cleaned = 0;

  for (const [id, session] of sessions) {
    const age = now - session.lastAccessed.getTime();
    if (age > SESSION_EXPIRY_MS) {
      sessions.delete(id);
      cleaned++;
    }
  }

  return cleaned;
}

/**
 * Get the number of active sessions.
 */
export function getSessionCount(): number {
  return sessions.size;
}

/**
 * Clear all sessions (useful for testing).
 */
export function clearAllSessions(): void {
  sessions.clear();
}
